src/IntegralArrays.jl

```diff
@@ -1,5 +1,73 @@
 module IntegralArrays
 
 # package code goes here
+using Images, IntervalSets
+
+import Base: size, getindex, linearindexing
+
+export IntegralArray
+
+"""
+```
+integral_array = IntegralArray(A)
+```
+
+Returns the integral array of an array. The integral array is calculated by assigning
+to each cell the sum of all cells above it and to its left, i.e. the rectangle from
+(1, 1) to the pixel. An integral array is a data structure which helps in efficient 
+calculation of sum of pixels in a rectangular subset of an array. 
+
+```
+sum = integral_array[ytop..ybot, xtop..xbot]
+sum = integral_array[y ± σy, x ± σx]
+sum = integral_array[CartesianIndex(y, x) ± σ]
+```
+
+The sum of a window in an array can be directly calculated using four array references of 
+the integral array, irrespective of the size of the window, given the `yrange` and `xrange` 
+of the window. Given an integral array - 
+        
+        A - - - - - - B -
+        - * * * * * * * -
+        - * * * * * * * -
+        - * * * * * * * -
+        - * * * * * * * -
+        - * * * * * * * -
+        C * * * * * * D - 
+        - - - - - - - - -
+
+The sum of pixels in the area denoted by * is given by S = D + A - B - C.
+"""
+immutable IntegralArray{T, N, A} <: AbstractArray{T, N}
+    data::A
+end
+
+function IntegralArray(array::AbstractArray)
+    integral_array = Array{Images.accum(eltype(array))}(size(array))
+    sd = coords_spatial(array)
+    cumsum!(integral_array, array, sd[1])
+    for i = 2:length(sd)
+        cumsum!(integral_array, integral_array, sd[i])
+    end
+    IntegralArray{eltype(array), ndims(array), typeof(array)}(integral_array)
+end
+
+linearindexing(A::IntegralArray) = Base.LinearFast()
+size(A::IntegralArray) = size(A.data)
+getindex(A::IntegralArray, i::Int...) = A.data[i...]
+
+getindex(A::IntegralArray, ids::Tuple...) = getindex(A, ids[1]...)
+
+function getindex(A::IntegralArray, i::ClosedInterval...)
```

Comment:
Best to declare this with two indexes, `i` and `j`, since you (1) require 2 in the implementation, and (2) discard more than 2. Once we generalize it to higher dimensions, it will become
```julia
getindex{T,N}(A::IntegralArray{T,N}, i::Vararg{ClosedInterval,N})
```
which will require that the number of indices matches the dimensionality of the array. (This is Julia PR #11242.)